File: packages/typescript/src/FileSystem.ts
/**
 * The Daytona SDK provides comprehensive file system operations through the `fs` module in Sandboxes.
 * You can perform various operations like listing files, creating directories, reading and writing files, and more.
 * This guide covers all available file system operations and best practices.
 * 
 * @module FileSystem
 * 
 * @example
 * // Basic file operations
 * // Create a workspace
 * const workspace = await daytona.create();
 * 
 * // Create a directory
 * await workspace.fs.createFolder('/workspace/data', '755');
 * 
 * // Upload a file
 * const fileContent = new File(['content'], 'local_file.txt');
 * await workspace.fs.uploadFile('/workspace/data/file.txt', fileContent);
 * 
 * // List directory contents
 * const files = await workspace.fs.listFiles('/workspace');
 * files.forEach(file => {
 *   console.log(`Name: ${file.name}`);
 *   console.log(`Is directory: ${file.isDir}`);
 *   console.log(`Size: ${file.size}`);
 *   console.log(`Modified: ${file.modTime}`);
 * });
 * 
 * // Search file contents
 * const matches = await workspace.fs.findFiles(
 *   '/workspace/src',
 *   'text-of-interest'
 * );
 * matches.forEach(match => {
 *   console.log(`Absolute file path: ${match.file}`);
 *   console.log(`Line number: ${match.line}`);
 *   console.log(`Line content: ${match.content}\n`);
 * });
 * 
 * @example
 * // File manipulation
 * // Move files
 * await workspace.fs.moveFiles(
 *   '/workspace/data/old.txt',
 *   '/workspace/data/new.txt'
 * );
 * 
 * // Replace text in files
 * const results = await workspace.fs.replaceInFiles(
 *   ['/workspace/data/new.txt'],
 *   'old_version',
 *   'new_version'
 * );
 * 
 * // Set permissions
 * await workspace.fs.setFilePermissions(
 *   '/workspace/data/script.sh',
 *   {
 *     mode: '755',
 *     owner: 'daytona'
 *   }
 * );
 */

import {
  FileInfo,
  Match,
  ReplaceRequest,
  ReplaceResult,
  SearchFilesResponse,
  ToolboxApi,
} from '@daytonaio/api-client'
import { WorkspaceInstance } from './Workspace'

/**
 * Parameters for setting file permissions in the Sandbox.
 * 
 * @interface FilePermissionsParams
 * @property {string} [mode] - File mode/permissions in octal format (e.g. "644")
 * @property {string} [owner] - User owner of the file
 * @property {string} [group] - Group owner of the file
 * 
 * @example
 * const permissions: FilePermissionsParams = {
 *   mode: '644',
 *   owner: 'daytona',
 *   group: 'users'
 * };
 */
export type FilePermissionsParams = {
  /** Group owner of the file */
  group?: string
  /** File mode/permissions in octal format (e.g. "644") */
  mode?: string
  /** User owner of the file */
  owner?: string
}

/**
 * Provides file system operations within a Sandbox.
 * 
 * This class implements a high-level interface to file system operations that can
 * be performed within a Daytona Sandbox. It supports common operations like
 * creating, deleting, and moving files, as well as searching file contents and
 * managing permissions.
 */
export class FileSystem {
  constructor(
    private readonly instance: WorkspaceInstance,
    private readonly toolboxApi: ToolboxApi,
  ) {}

  /**
   * Create a new directory in the Sandbox with specified permissions.
   * 
   * @param {string} path - Path where the directory should be created
   * @param {string} mode - Directory permissions in octal format (e.g. "755")
   * @returns {Promise<void>}
   * 
   * @example
   * // Create a directory with standard permissions
   * await fs.createFolder('/app/data', '755');
   */
  public async createFolder(path: string, mode: string): Promise<void> {
    const response = await this.toolboxApi.createFolder(this.instance.id, path, mode)
    return response.data
  }

  /**
   * Deletes a file or directory from the Sandbox.
   * 
   * @param {string} path - Path to the file or directory to delete
   * @returns {Promise<void>}
   * 
   * @example
   * // Delete a file
   * await fs.deleteFile('/app/temp.log');
   */
  public async deleteFile(path: string): Promise<void> {
    const response = await this.toolboxApi.deleteFile(this.instance.id, path)
    return response.data
  }

  /**
   * Downloads a file from the Sandbox.
   * 
   * @param {string} path - Path to the file to download
   * @returns {Promise<Blob>} The file contents as a Blob
   * 
   * @example
   * // Download and process a file
   * const fileBlob = await fs.downloadFile('/app/data.json');
   * console.log('File content:', fileBlob.toString());
   */
  public async downloadFile(path: string): Promise<Blob> {
    const response = await this.toolboxApi.downloadFile(this.instance.id, path)
    return response.data
  }

  /**
   * Searches for text patterns within files in the Sandbox.
   * 
   * @param {string} path - Directory to search in
   * @param {string} pattern - Search pattern
   * @returns {Promise<Array<Match>>} Array of matches with file and line information
   * 
   * @example
   * // Find all TODO comments in TypeScript files
   * const matches = await fs.findFiles('/app/src', 'TODO:');
   * matches.forEach(match => {
   *   console.log(`${match.file}:${match.line}: ${match.content}`);
   * });
   */
  public async findFiles(path: string, pattern: string): Promise<Array<Match>> {
    const response = await this.toolboxApi.findInFiles(this.instance.id, path, pattern)
    return response.data
  }

  /**
   * Retrieves detailed information about a file or directory.
   * 
   * @param {string} path - Path to the file or directory
   * @returns {Promise<FileInfo>} Detailed file information including size, permissions, modification time
   * 
   * @example
   * // Get file details
   * const info = await fs.getFileDetails('/app/config.json');
   * console.log(`Size: ${info.size}, Modified: ${info.modTime}`);
   */
  public async getFileDetails(path: string): Promise<FileInfo> {
    const response = await this.toolboxApi.getFileInfo(this.instance.id, path)
    return response.data
  }

  /**
   * Lists contents of a directory in the Sandbox.
   * 
   * @param {string} path - Directory path to list
   * @returns {Promise<FileInfo[]>} Array of file and directory information
   * 
   * @example
   * // List directory contents
   * const files = await fs.listFiles('/app/src');
   * files.forEach(file => {
   *   console.log(`${file.name} (${file.size} bytes)`);
   * });
   */
  public async listFiles(path: string): Promise<FileInfo[]> {
    const response = await this.toolboxApi.listFiles(this.instance.id, path)
    return response.data
  }

  /**
   * Moves or renames a file or directory.
   * 
   * @param {string} source - Source path
   * @param {string} destination - Destination path
   * @returns {Promise<void>}
   * 
   * @example
   * // Move a file to a new location
   * await fs.moveFiles('/app/temp/data.json', '/app/data/data.json');
   */
  public async moveFiles(source: string, destination: string): Promise<void> {
    const response = await this.toolboxApi.moveFile(this.instance.id, source, destination)
    return response.data
  }

  /**
   * Replaces text content in multiple files.
   * 
   * @param {string[]} files - Array of file paths to process
   * @param {string} pattern - Pattern to replace
   * @param {string} newValue - Replacement text
   * @returns {Promise<Array<ReplaceResult>>} Results of the replace operation for each file
   * 
   * @example
   * // Update version number across multiple files
   * const results = await fs.replaceInFiles(
   *   ['/app/package.json', '/app/version.ts'],
   *   '"version": "1.0.0"',
   *   '"version": "1.1.0"'
   * );
   */
  public async replaceInFiles(
    files: string[],
    pattern: string,
    newValue: string,
  ): Promise<Array<ReplaceResult>> {
    const replaceRequest: ReplaceRequest = {
      files,
      newValue,
      pattern,
    }

    const response = await this.toolboxApi.replaceInFiles(this.instance.id, replaceRequest)
    return response.data
  }

  /**
   * Searches for files and directories by name pattern in the Sandbox.
   * 
   * @param {string} path - Directory to search in
   * @param {string} pattern - File name pattern (supports globs)
   * @returns {Promise<SearchFilesResponse>} Search results with matching files
   * 
   * @example
   * // Find all TypeScript files
   * const result = await fs.searchFiles('/app', '*.ts');
   * result.files.forEach(file => console.log(file));
   */
  public async searchFiles(
    path: string,
    pattern: string,
  ): Promise<SearchFilesResponse> {
    const response = await this.toolboxApi.searchFiles(this.instance.id, path, pattern)
    return response.data
  }

  /**
   * Sets permissions and ownership for a file or directory.
   * 
   * @param {string} path - Path to the file or directory
   * @param {FilePermissionsParams} permissions - Permission settings
   * @returns {Promise<void>}
   * 
   * @example
   * // Set file permissions and ownership
   * await fs.setFilePermissions('/app/script.sh', {
   *   owner: 'daytona',
   *   group: 'users',
   *   mode: '755'  // Execute permission for shell script
   * });
   */
  public async setFilePermissions(
    path: string,
    permissions: FilePermissionsParams,
  ): Promise<void> {
    const response = await this.toolboxApi.setFilePermissions(this.instance.id, path, permissions.owner!, permissions.group!, permissions.mode!)
    return response.data
  }

  /**
   * Uploads a file to the Sandbox.
   * 
   * @param {string} path - Destination path in the Sandbox
   * @param {File} file - File to upload
   * @returns {Promise<void>}
   * 
   * @example
   * // Upload a configuration file
   * const configFile = new File(['{"setting": "value"}'], 'config.json');
   * await fs.uploadFile('/app/config.json', configFile);
   */
  public async uploadFile(path: string, file: File): Promise<void> {
    const response = await this.toolboxApi.uploadFile(this.instance.id, path, file)
    return response.data
  }
}

File: packages/python/src/daytona_sdk/filesystem.py
"""
The Daytona SDK provides comprehensive file system operations through the `fs` module in Sandboxes.
You can perform various operations like listing files, creating directories, reading and writing files, and more.
This guide covers all available file system operations and best practices.

Examples:
    Basic file operations:
    ```python
    workspace = daytona.create()
    
    # Create a directory
    workspace.fs.create_folder("/workspace/data", "755")
    
    # Upload a file
    with open("local_file.txt", "rb") as f:
        content = f.read()
    workspace.fs.upload_file("/workspace/data/file.txt", content)
    
    # List directory contents
    files = workspace.fs.list_files("/workspace")
    for file in files:
        print(f"Name: {file.name}")
        print(f"Is directory: {file.is_dir}")
        print(f"Size: {file.size}")
        print(f"Modified: {file.mod_time}")
    
    # Search file contents
    matches = workspace.fs.find_files(
        path="/workspace/src",
        pattern="text-of-interest"
    )
    for match in matches:
        print(f"Absolute file path: {match.file}")
        print(f"Line number: {match.line}")
        print(f"Line content: {match.content}")
        print("\n")
    ```

    File manipulation:
    ```python
    # Move files
    workspace.fs.move_files(
        "/workspace/data/old.txt",
        "/workspace/data/new.txt"
    )
    
    # Replace text in files
    results = workspace.fs.replace_in_files(
        files=["/workspace/data/new.txt"],
        pattern="old_version",
        new_value="new_version"
    )
    
    # Set permissions
    workspace.fs.set_file_permissions(
        path="/workspace/data/script.sh",
        mode="755",
        owner="daytona"
    )
    ```

Note:
    All paths should be absolute paths within the Sandbox if not explicitly
    stated otherwise.
"""

from typing import List
from daytona_api_client import (
    FileInfo,
    Match,
    ReplaceRequest,
    ReplaceResult,
    SearchFilesResponse,
    ToolboxApi,
)
from daytona_sdk._utils.errors import intercept_errors
from .protocols import WorkspaceInstance


class FileSystem:
    """Provides file system operations within a Sandbox.

    This class implements a high-level interface to file system operations that can
    be performed within a Daytona Sandbox. It supports common operations like
    creating, deleting, and moving files, as well as searching file contents and
    managing permissions.

    Attributes:
        instance (WorkspaceInstance): The Sandbox instance this file system belongs to.
    """

    def __init__(self, instance: WorkspaceInstance, toolbox_api: ToolboxApi):
        """Initializes a new FileSystem instance.

        Args:
            instance (WorkspaceInstance): The Sandbox instance this file system belongs to.
            toolbox_api (ToolboxApi): API client for Sandbox operations.
        """
        self.instance = instance
        self.toolbox_api = toolbox_api

    @intercept_errors(message_prefix="Failed to create folder: ")
    def create_folder(self, path: str, mode: str) -> None:
        """Creates a new directory in the Sandbox.

        This method creates a new directory at the specified path with the given
        permissions.

        Args:
            path (str): Absolute path where the folder should be created.
            mode (str): Folder permissions in octal format (e.g., "755" for rwxr-xr-x).

        Example:
            ```python
            # Create a directory with standard permissions
            workspace.fs.create_folder("/workspace/data", "755")

            # Create a private directory
            workspace.fs.create_folder("/workspace/secrets", "700")
            ```
        """
        self.toolbox_api.create_folder(
            workspace_id=self.instance.id, path=path, mode=mode
        )

    @intercept_errors(message_prefix="Failed to delete file: ")
    def delete_file(self, path: str) -> None:
        """Deletes a file from the Sandbox.

        This method permanently deletes a file from the Sandbox.

        Args:
            path (str): Absolute path to the file to delete.

        Example:
            ```python
            # Delete a file
            workspace.fs.delete_file("/workspace/data/old_file.txt")
            ```
        """
        self.toolbox_api.delete_file(
            workspace_id=self.instance.id, path=path
        )

    @intercept_errors(message_prefix="Failed to download file: ")
    def download_file(self, path: str) -> bytes:
        """Downloads a file from the Sandbox.

        This method retrieves the contents of a file from the Sandbox.

        Args:
            path (str): Absolute path to the file to download.

        Returns:
            bytes: The file contents as a bytes object.

        Example:
            ```python
            # Download and save a file locally
            content = workspace.fs.download_file("/workspace/data/file.txt")
            with open("local_copy.txt", "wb") as f:
                f.write(content)

            # Download and process text content
            content = workspace.fs.download_file("/workspace/data/config.json")
            config = json.loads(content.decode('utf-8'))
            ```
        """
        return self.toolbox_api.download_file(
            workspace_id=self.instance.id, path=path
        )

    @intercept_errors(message_prefix="Failed to find files: ")
    def find_files(self, path: str, pattern: str) -> List[Match]:
        """Searches for files containing a pattern.

        This method searches file contents for a specified pattern, similar to
        the grep command.

        Args:
            path (str): Absolute path to the file or directory to search. If the path is a directory, the search will be performed recursively.
            pattern (str): Search pattern to match against file contents.

        Returns:
            List[Match]: List of matches found in files. Each Match object includes:
                - file: Path to the file containing the match
                - line: The line number where the match was found
                - content: The matching line content

        Example:
            ```python
            # Search for TODOs in Python files
            matches = workspace.fs.find_files("/workspace/src", "TODO:")
            for match in matches:
                print(f"{match.file}:{match.line}: {match.content.strip()}")
            ```
        """
        return self.toolbox_api.find_in_files(
            workspace_id=self.instance.id, path=path, pattern=pattern
        )

    @intercept_errors(message_prefix="Failed to get file info: ")
    def get_file_info(self, path: str) -> FileInfo:
        """Gets detailed information about a file.

        This method retrieves metadata about a file or directory, including its
        size, permissions, and timestamps.

        Args:
            path (str): Absolute path to the file or directory.

        Returns:
            FileInfo: Detailed file information including:
                - name: File name
                - is_dir: Whether the path is a directory
                - size: File size in bytes
                - mode: File permissions
                - mod_time: Last modification timestamp
                - permissions: File permissions in octal format
                - owner: File owner
                - group: File group

        Example:
            ```python
            # Get file metadata
            info = workspace.fs.get_file_info("/workspace/data/file.txt")
            print(f"Size: {info.size} bytes")
            print(f"Modified: {info.mod_time}")
            print(f"Mode: {info.mode}")

            # Check if path is a directory
            info = workspace.fs.get_file_info("/workspace/data")
            if info.is_dir:
                print("Path is a directory")
            ```
        """
        return self.toolbox_api.get_file_info(
            workspace_id=self.instance.id, path=path
        )

    @intercept_errors(message_prefix="Failed to list files: ")
    def list_files(self, path: str) -> List[FileInfo]:
        """Lists files and directories in a given path.

        This method returns information about all files and directories in the
        specified directory, similar to the ls -l command.

        Args:
            path (str): Absolute path to the directory to list contents from.

        Returns:
            List[FileInfo]: List of file and directory information. Each FileInfo
                object includes the same fields as described in get_file_info().

        Example:
            ```python
            # List directory contents
            files = workspace.fs.list_files("/workspace/data")

            # Print files and their sizes
            for file in files:
                if not file.is_dir:
                    print(f"{file.name}: {file.size} bytes")

            # List only directories
            dirs = [f for f in files if f.is_dir]
            print("Subdirectories:", ", ".join(d.name for d in dirs))
            ```
        """
        return self.toolbox_api.list_files(
            workspace_id=self.instance.id, path=path
        )

    @intercept_errors(message_prefix="Failed to move files: ")
    def move_files(self, source: str, destination: str) -> None:
        """Moves files from one location to another.

        This method moves or renames a file or directory. The parent directory
        of the destination must exist.

        Args:
            source (str): Absolute path to the source file or directory.
            destination (str): Absolute path to the destination.

        Example:
            ```python
            # Rename a file
            workspace.fs.move_files(
                "/workspace/data/old_name.txt",
                "/workspace/data/new_name.txt"
            )

            # Move a file to a different directory
            workspace.fs.move_files(
                "/workspace/data/file.txt",
                "/workspace/archive/file.txt"
            )

            # Move a directory
            workspace.fs.move_files(
                "/workspace/old_dir",
                "/workspace/new_dir"
            )
            ```
        """
        self.toolbox_api.move_file(
            workspace_id=self.instance.id,
            source=source,
            destination=destination,
        )

    @intercept_errors(message_prefix="Failed to replace in files: ")
    def replace_in_files(
        self, files: List[str], pattern: str, new_value: str
    ) -> List[ReplaceResult]:
        """Replaces text in multiple files.

        This method performs search and replace operations across multiple files.

        Args:
            files (List[str]): List of absolute file paths to perform replacements in.
            pattern (str): Pattern to search for.
            new_value (str): Text to replace matches with.

        Returns:
            List[ReplaceResult]: List of results indicating replacements made in
                each file. Each ReplaceResult includes:
                - file: Path to the modified file
                - success: Whether the operation was successful
                - error: Error message if the operation failed

        Example:
            ```python
            # Replace in specific files
            results = workspace.fs.replace_in_files(
                files=["/workspace/src/file1.py", "/workspace/src/file2.py"],
                pattern="old_function",
                new_value="new_function"
            )

            # Print results
            for result in results:
                if result.success:
                    print(f"{result.file}: {result.success}")
                else:
                    print(f"{result.file}: {result.error}")
            ```
        """
        replace_request = ReplaceRequest(
            files=files, new_value=new_value, pattern=pattern
        )

        return self.toolbox_api.replace_in_files(
            workspace_id=self.instance.id, replace_request=replace_request
        )

    @intercept_errors(message_prefix="Failed to search files: ")
    def search_files(self, path: str, pattern: str) -> SearchFilesResponse:
        """Searches for files and directories matching a pattern in their names.

        This method searches for files and directories whose names match the
        specified pattern. The pattern can be a simple string or a glob pattern.

        Args:
            path (str): Absolute path to the root directory to start search from.
            pattern (str): Pattern to match against file names. Supports glob
                patterns (e.g., "*.py" for Python files).

        Returns:
            SearchFilesResponse: Search results containing:
                - files: List of matching file and directory paths

        Example:
            ```python
            # Find all Python files
            result = workspace.fs.search_files("/workspace", "*.py")
            for file in result.files:
                print(file)

            # Find files with specific prefix
            result = workspace.fs.search_files("/workspace/data", "test_*")
            print(f"Found {len(result.files)} test files")
            ```
        """
        return self.toolbox_api.search_files(
            workspace_id=self.instance.id, path=path, pattern=pattern
        )

    @intercept_errors(message_prefix="Failed to set file permissions: ")
    def set_file_permissions(
        self, path: str, mode: str = None, owner: str = None, group: str = None
    ) -> None:
        """Sets permissions and ownership for a file or directory.

        This method allows changing the permissions and ownership of a file or
        directory. Any of the parameters can be None to leave that attribute
        unchanged.

        Args:
            path (str): Absolute path to the file or directory.
            mode (Optional[str]): File mode/permissions in octal format
                (e.g., "644" for rw-r--r--).
            owner (Optional[str]): User owner of the file.
            group (Optional[str]): Group owner of the file.

        Example:
            ```python
            # Make a file executable
            workspace.fs.set_file_permissions(
                path="/workspace/scripts/run.sh",
                mode="755"  # rwxr-xr-x
            )

            # Change file owner
            workspace.fs.set_file_permissions(
                path="/workspace/data/file.txt",
                owner="daytona",
                group="daytona"
            )
            ```
        """
        self.toolbox_api.set_file_permissions(
            workspace_id=self.instance.id,
            path=path,
            mode=mode,
            owner=owner,
            group=group,
        )

    @intercept_errors(message_prefix="Failed to upload file: ")
    def upload_file(self, path: str, file: bytes) -> None:
        """Uploads a file to the Sandbox.

        This method uploads a file to the specified path in the Sandbox. The
        parent directory must exist. If a file already exists at the destination
        path, it will be overwritten.

        Args:
            path (str): Absolute destination path in the Sandbox.
            file (bytes): File contents as a bytes object.

        Example:
            ```python
            # Upload a text file
            content = b"Hello, World!"
            workspace.fs.upload_file("/workspace/data/hello.txt", content)

            # Upload a local file
            with open("local_file.txt", "rb") as f:
                content = f.read()
            workspace.fs.upload_file("/workspace/data/file.txt", content)

            # Upload binary data
            import json
            data = {"key": "value"}
            content = json.dumps(data).encode('utf-8')
            workspace.fs.upload_file("/workspace/data/config.json", content)
            ```
        """
        self.toolbox_api.upload_file(
            workspace_id=self.instance.id, path=path, file=file
        )
